File: solutions/cpp/grains/1/grains.cpp
#include "grains.h"

namespace grains {
    auto square(const uint8_t& input) -> uint64_t {
        if (input < 1 or input > 64) {
            throw(std::domain_error("invalid argument"));
        }

        const auto index = input - 1;
        
        return board[index];
    }

    auto total() -> uint64_t {
        return sum;
    }
}  // namespace grains

File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n
 | (n < 1) || (n > 64) = Nothing
 | otherwise           = Just (2 ^ (n - 1))

total :: Integer
-- sum of a geometric progression from 2^0 to 2^64
total = gpSum 1 2 64

gpSum :: Integer -> Integer -> Integer -> Integer
gpSum a r n = div (a * (r ^ n - 1)) (r - 1)
